jena-arq/src/main/java/org/apache/jena/sparql/exec/http/QueryExecHTTP.java

```diff
@@ -190,10 +191,20 @@ private RowSet execRowSet() {
 
         // Map to lang, with pragmatic alternatives.
         Lang lang = WebContent.contentTypeToLangResultSet(actualContentType);
-        if ( lang == null )
-            throw new QueryException("Endpoint returned Content-Type: " + actualContentType + " which is not recognized for SELECT queries");
-        if ( !ResultSetReaderRegistry.isRegistered(lang) )
-            throw new QueryException("Endpoint returned Content-Type: " + actualContentType + " which is not supported for SELECT queries");
+        boolean unknownLang = lang == null;
+        boolean unsupportedFormat = !unknownLang && !ResultSetReaderRegistry.isRegistered(lang);
+
+        if ( unknownLang || unsupportedFormat ) {
+            int statusCode = response.statusCode();
+            String statusCodeMsg = HttpSC.getMessage(statusCode);
+            String bodyStr = IO.abbreviateAsUTF8(in, 1024, "...");
+            String errorTerm = unknownLang ? "recognized" : "supported";
+
+            throw new QueryException(String.format(
+                    "Endpoint returned Content-Type: %s which is not %s for SELECT queries. Status code %d %s, body: %s",
+                    actualContentType, errorTerm, statusCode, statusCodeMsg, bodyStr));
```

Comment:
Would it be worth dumping the whole body into a string and putting that in an inner exception, or creating a new derived subclass of `QueryException`?

In your example the abbreviated content isn't particularly enlightening so having the full response available somewhere for advanced users to further inspect might be useful

Reply:
In the code the limit is hard coded to 1024 characters; may example only showed an excerpt of the excerpt. Also, the threshold could be made configurable via the context.
What I want to avoid is unconditionally loading everything into a string because for large incorrect responses this could again make investigation difficult.

I was also thinking that QueryExceptionHTTP would be more appropriate - but the original code used QueryExeception; so I thought maybe @afs had something in mind when he did not use the former (such as that a non-handleable response content type is not really an HTTP issue).

Furthermore, there seems to be more strange things going on with Jena over Virtuoso. In my specific case the problem is that Virtuoso sometimes returns plain text/html for queries - rather than some virtuoso error. But with this PR at least this becomes visible. So it's seems to be related to that Virtuoso instance (DBpedia) rather than Jena.

I also intend to extend this PR with logging the HTTP request headers on error during this week - because some of my queries sometimes hang - but without knowing the exact headers I can't reproduce it - if I paste the query into the Web form it works.

Reply:
In practice, I think developers will often check the endpoint with tools such as curl and wget to get more information. The whole thing might be very large for a log item.

(and, yes, Virtuoso does seem to return text/html - it seems to snoop the request to determine whether it is a programmatic query or now, and also users don't always find its endpoint but the SPARQL web UI.)

> QueryExectipionHTTP

I _think_ it is because it is not strictly an HTTP-level error. The HTTP mechanisms are fine. It's the application usage that is mismatched.

Reply:
> I think developers will often check the endpoint with tools such as curl and wget to get more information. 
That's a chicken & egg problem:  Sometimes queries fail for no apparent reason - and because Jena so far does not log details about failed request it's nearly impossible to reproduce the exact requests with curl and wget for further investigation.

> I think it is because it is not strictly an HTTP-level error.

Yes, that's what I thought was the reason - the question is whether reuse of QueryExceptionHTTP would still be justified because its still a QueryException *related* to an HTTP request. Also, that exception has a response field that (IIRC) does not show on the stack-trace but can be printed out on demand.

Reply:
The response code is 200 to get to this point in the code. 3xx, 4xx and 5xx are already handled elsewhere.